nft_ingester/src/program_transformers/token/mod.rs

```diff
@@ -121,20 +121,38 @@ pub async fn handle_token_program_account<'a, 'b, 'c>(
                 )
                 .build(DbBackend::Postgres);
             query.sql = format!(
-                "{} WHERE excluded.slot_updated > tokens.slot_updated",
+                "{} WHERE excluded.slot_updated >= tokens.slot_updated",
                 query.sql
             );
             db.execute(query).await?;
+
             let asset_update: Option<asset::Model> = asset::Entity::find_by_id(key_bytes.clone())
-                .filter(asset::Column::OwnerType.eq("single"))
+                .filter(
+                    asset::Column::OwnerType
+                        .eq(OwnerType::Single)
+                        .or(asset::Column::OwnerType
+                            .eq(OwnerType::Unknown)
+                            .and(asset::Column::Supply.eq(1))),
+                )
                 .one(db)
                 .await?;
             if let Some(asset) = asset_update {
-                let mut active: asset::ActiveModel = asset.into();
+                let mut active: asset::ActiveModel = asset.clone().into();
                 active.supply = Set(m.supply as i64);
                 active.supply_mint = Set(Some(key_bytes));
+
+                // Update owner_type based on the supply.
+                if asset.owner_type == OwnerType::Unknown {
+                    active.owner_type = match m.supply.cmp(&1) {
+                        std::cmp::Ordering::Equal => Set(OwnerType::Single),
+                        std::cmp::Ordering::Greater => Set(OwnerType::Token),
+                        _ => NotSet,
+                    }
+                }
+
                 active.save(db).await?;
```

Comment:
Since we've recently done all these studies for race conditions due to concurrent processes, one pattern I keep an eye out for and flag as potential issue is the one where there's a separate read and write SQL statements, which is happening here.

I'm actually still ramping up on some of the token indexing, but is this scenario possible?:
1. First process: we are indexing a token with supply of 1, and one record is returned by the `find_by_id` (`OwnerType::Unknown` and `Column::Supply.eq(1)`),
2. Second (concurent) process indexing after token is burned, sets supply to 0.
3. First process would erroneously set token back to supply of 1 with the `active.save`.

Should we consider simply combining these SQL statements?

Reply:
You're bang on. We should avoid doing read+write combos. One of the next tasks in our backlog is to fix the race conditions and re-work the NFT indexing, but its a much bigger and riskier project. 

The scope of this PR is to simply improve on the existing setup by adding a few retries to work around some race conditions. Also, aligning the code w/ the Helius setup will make it easier for this repo to benefit from our race condition fixes.